src/libstore/gc.cc

```diff
@@ -330,18 +335,184 @@ Roots LocalStore::findRoots()
+                fdDir.close();
+
+                auto mapFd = AutoCloseFD{openat(procFd.get(), "maps", O_RDONLY | O_CLOEXEC)};
+                if (!mapFd) {
+                    if (errno == ENOENT || errno == EACCES)
+                        continue;
+                    throw SysError(format("opening /proc/%1%/maps") % ent->d_name);
+                }
+                while (true) {
+                    try {
+                        auto line = readLine(mapFd.get());
+                        auto match = std::smatch{};
+                        if (std::regex_match(line, match, mapRegex))
+                            paths.emplace(match[1]);
+                    } catch (EndOfFile & e) {
+                        break;
+                    }
+                }
+                mapFd = -1;
 
-    debug(format("executing ‘%1%’ to find additional roots") % rootFinder);
+                auto envFd = AutoCloseFD{openat(procFd.get(), "environ", O_RDONLY | O_CLOEXEC)};
+                if (!envFd) {
+                    if (errno == ENOENT || errno == EACCES)
+                        continue;
+                    throw SysError(format("opening /proc/%1%/environ") % ent->d_name);
+                }
+                auto envString = readFile(envFd.get());
+                envFd = -1;
+                auto env_end = std::sregex_iterator{};
+                for (auto i = std::sregex_iterator{envString.begin(), envString.end(), storePathRegex}; i != env_end; ++i)
+                    paths.emplace(i->str());
+            }
+        }
+        if (errno)
+            throw SysError("iterating /proc");
+    }
 
-    string result = runProgram(rootFinder);
+    auto dev_null = AutoCloseFD{open("/dev/null", O_RDWR | O_CLOEXEC)};
+    Pipe lsof_out, lsof_report;
+    lsof_out.create();
+    lsof_report.create();
+    auto lsof_pid = Pid(fork());
+    switch (lsof_pid) {
+        case -1:
+            throw SysError("forking to run lsof");
+        case 0:
+            if (dup2(dev_null.get(), STDERR_FILENO) == -1)
+                goto err;
+            if (dup2(dev_null.get(), STDIN_FILENO) == -1)
+                goto err;
+            if (dup2(lsof_out.writeSide.get(), STDOUT_FILENO) == -1)
+                goto err;
+            execlp("lsof", "lsof", "-n", "-w", "-F", "n", NULL);
+err:
+            write(lsof_report.writeSide.get(), &errno, sizeof errno);
+            _exit(1);
+    }
+    dev_null = -1;
+    lsof_out.writeSide = -1;
+    lsof_report.writeSide = -1;
+    std::remove_reference<decltype(errno)>::type child_errno;
+    auto report_read_count = size_t{0};
+    while (report_read_count < sizeof child_errno) {
+        auto res = read(lsof_report.readSide.get(), reinterpret_cast<unsigned char *>(&child_errno) + report_read_count, sizeof child_errno - report_read_count);
+        if (res == -1)
+            throw SysError("reading from lsof report pipe");
+        if (res == 0)
+            break;
+        report_read_count += res;
+    }
```

Comment:
Can't put stderr to dev null, can't differentiate error in lsof (which is a problem) from lsof not being installed.